src/Analyzers/Core/Analyzers/NamespaceMatchFolder/AbstractNamespaceMatchFolderDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,196 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Analyzers.NamespaceSync
+{
+    internal abstract class AbstractNamespaceMatchFolderDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        public const string RootNamespaceOption = "build_property.RootNamespace";
+        public const string ProjectDirOption = "build_property.ProjectDir";
+        public const string TargetNamespace = "TargetNamespace";
+
+        private static readonly LocalizableResourceString s_localizableTitle = new(
+          nameof(AnalyzersResources.Namespace_does_not_match_folder_structure), AnalyzersResources.ResourceManager, typeof(AnalyzersResources));
+
+        private static readonly LocalizableResourceString s_localizableInsideMessage = new(
+            nameof(AnalyzersResources.Namespace_0_does_not_match_folder_structure_expected_1), AnalyzersResources.ResourceManager, typeof(AnalyzersResources));
+
+        protected AbstractNamespaceMatchFolderDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.NamespaceSyncAnalyzerDiagnosticId,
+                EnforceOnBuildValues.NamespaceSync,
+                CodeStyleOptions2.PreferNamespaceMatchFolderStructure,
+                s_localizableTitle,
+                s_localizableInsideMessage)
+        {
+        }
+    }
+
+    internal abstract class AbstractNamespaceMatchFolderDiagnosticAnalyzer<TNamespaceSyntax> : AbstractNamespaceMatchFolderDiagnosticAnalyzer
+        where TNamespaceSyntax : SyntaxNode
+    {
+        private static readonly SymbolDisplayFormat s_namespaceDisplayFormat = SymbolDisplayFormat
+            .FullyQualifiedFormat
+            .WithGlobalNamespaceStyle(SymbolDisplayGlobalNamespaceStyle.Omitted);
+
+        /// <summary>
+        /// Gets the language specific syntax facts
+        /// </summary>
```

Comment:
nit: i'm always on the fence about these sorts of commetns.  to me a comment has to 'earn' its place in teh code.  meaning that it needs to provide enough actual value/meaning to warrant the 3 lines added here.  In this case, we don't have that.  `ISyntaxFacts GetSyntaxFacts();` is totally clear on what it is doing, and certainly already indicates that it "gets the language specific syntax facts". So i would remove this.